ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -197,52 +216,83 @@ densityDisconnect ::
+            (NotOrigin cand, Just latest)
+              | latest > cand -> LatestSlotForecast latest
+              | otherwise -> LatestSlotCandidate cand
+
+          -- If the peer has more headers that it hasn't sent yet, each slot between the latest header we know of and
+          -- the end of the Genesis window could contain a block, so the upper bound for the total number of blocks in
+          -- the window is the sum of the known blocks and that number of remaining slots.
+          -- If the slot of the latest header we know of is _after_ the end of the Genesis window (either because the
+          -- candidate fragment extends beyond it or because we are waiting to validate a header beyond the forecast
+          -- horizon that we already received), there can be no headers in between and 'potentialSlots' is 0.
+          SlotNo potentialSlots
+            | caughtUp
+            = 0
             | otherwise
-            = AF.length candidateSuffix >= fromIntegral k
-          lastSlot | isJust maybeForecast, lastSlotForecast >= lastSlotCandidate = Left lastSlotForecast
-                   | otherwise = Right lastSlotCandidate
-      pure (peer, DensityBounds {fragment, offersMoreThanK, lowerBound, upperBound, hasBlockAfter = hasBlockAfterGenesisWindow, lastSlot})
+            = case latestSlot of
+              -- The peer either has not advertised its tip yet or simply has no blocks whatsoever and won't progress
+              -- either.
+              -- In the latter case, it should be killed by the LoP.
+              -- REVIEW: What do we want to do in tests here? Both cases are possible, and we cannot distinguish them
+              -- without the tip (even with the tip we have been relying on the invariant that each peer advertises the
+              -- tip before any of them sends headers).
+              -- NoLatestSlot -> 0
+              NoLatestSlot -> SlotNo sgen
+              -- 'endOfGenesisWindow' is exclusive, so we have to add 1 to the non-exclusive last slot
+              LatestSlotCandidate slot -> intervalLength endOfGenesisWindow (slot + 1)
+              -- If the candidate fragment's last slot is smaller than the slot set after receiving a header, we are
+              -- stuck at the forecast horizon, which implies that there is a header after the Genesis window.
+              LatestSlotForecast _ -> 0
+
+          -- The number of blocks within the Genesis window we know with certainty
+          lowerBound = fromIntegral $ AF.length fragment
+
+          upperBound = lowerBound + potentialSlots
+
+          -- The number of blocks we know of on the candidate chain, not limited to the Genesis window.
+          totalBlockCount = fromIntegral (AF.length candidateSuffix)
+
+          -- Does the peer have more than k known blocks in _total_ after the intersection?
+          -- If not, it is not qualified to compete by density (yet).
+          offersMoreThanK = case latestSlot of
```

Comment:
I'm not sure this plays well with lightweight checkpointing, where the honest chain might have lower densities. Why are peers required to offer more than k blocks?

Reply:
FTR: When almost caught-up, there might be light disagreements (eg forks of depth 1) even between honest peers due to normal Praos dynamics. Without this criterion, we might disconnect from a peer because it offers just one block after the intersection, compared to two for another peer.

Lightweight checkpointing should be orthogonal, in case of a disaster scenario, we just have to add sufficient checkpoints such that any possible alternative chain is invalid.

Reply:
In that case, we should restrict the computation of `offersMoreThanK` to the actual headers that have been validated, and not increase it speculatively depending on headers that are after the forecast horizon and therefore weren't checked for checkpoints.


Also, the forecast horizon isn't relevant to the scenario where `offersMoreThanK` is useful.

Reply:
Both Forecast and Candidate cases now compute the same `totalBlockCount>k`.